[Feat] 8주차 Spring Security 코드 구현

jang-namu opened the pull request:
## 📋 이슈 번호
close #111 
## 🛠 구현 사항

## 📚 기타

jang-namu committed:
feat: logging(postUser) 추가
slf4j + logback

jang-namu committed:
[Feat] logging(postUser) 추가
slf4j + logback

jang-namu committed:
[Feat] Authority 클래스 작성
User - Authority : One to Many 연관관계 설정

jang-namu committed:
[Feat] CustomUserDetails class
UserDetails implementation

jang-namu committed:
[Feat] CustomUserDetailSerivce class
UserDetailService implementation

jang-namu committed:
[Feat] SecurityConfig PasswordEncoder Bean 추가

jang-namu committed:
[Feat] JwtProvider class
jwt 생성 및 expirtation 검증, 토큰 추출, Authentication 추출

jang-namu committed:
[Feat] JwtAuthenticationFilter class
JWT를 이용한 인증을 수행하는 Filter
OncePerRequest 상속 -> 요청마다 한 번 수행
SecurityContextHolder에 Authentication 정보를 저장

jang-namu committed:
[Feat] SecurityConfig filterChain 메서드 작성

jang-namu committed:
feat: UserAccountService & UserAccountController
JWT 기반 회원가입, 로그인, 회원탈퇴 기능

jang-namu committed:
[Fix] secret 변경(토이프로젝트용으로 노출)
application.properties를 올리면 안 됨

jang-namu committed:
Merge branch 'chulhee' of https://github.com/inu-appcenter/server-study-15.5 into chulhee

이전 commit message 수정으로 인한 branch 정리

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/config/SecurityConfig.java:
```diff
@@ -1,37 +1,120 @@
+                */
+                .csrf().disable()
+                // CORS 설정
+                .cors(c -> {
```
향후에 설정이 많아지는 경우에는 corsFilter 등으로 따로 빼서 관리하는 게 좋을 것 같아!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/config/SecurityConfig.java:
```diff
@@ -1,37 +1,120 @@
+                        "/swagger-ui.html",
+                        "/webjars/**").permitAll()
+                .anyRequest().denyAll()
+                .and()
```
여기 이후에 나오는 AccessDeniedHandler / AuthenticationEntryPoint도 따로 관리하는 게 좋아보여!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/security/JwtProvider.java:
```diff
@@ -0,0 +1,94 @@
+    private final long expirationTime = 1000L * 60 * 60;
+
+    @PostConstruct
+    protected void init() {
```
PostConstruct를 사용한 이유는?
그리고 이게 필수적일 지?

jang-namu replied:
음.. 아무래도 seed라고 할 수 있는 Secret과 SecretKey를 분리해서 작성하다 보니 쓰게됐어.
secret의 경우 @Value 어노테이션을 통해 application.properties에 저장된 값을 가져오도록 작성했고
secretKey는 이를 이용해야하기 때문에 런타임에 secret 값이 불러와진 다음 init 메소드가 실행되어야 한다고 봤기 때문에 @PostConstruct를 사용했게 됐어.

더 좋은 방법이 있을까?

Juser0 replied:
> 음.. 아무래도 seed라고 할 수 있는 Secret과 SecretKey를 분리해서 작성하다 보니 쓰게됐어. secret의 경우 @value 어노테이션을 통해 application.properties에 저장된 값을 가져오도록 작성했고 secretKey는 이를 이용해야하기 때문에 런타임에 secret 값이 불러와진 다음 init 메소드가 실행되어야 한다고 봤기 때문에 @PostConstruct를 사용했게 됐어.
> 
> 더 좋은 방법이 있을까?

나도 이 방법이 맞다고 생각해.
원정이형은 Key type으로 HMAC 암호화를 이야기했지만,
형처럼 value 어노테이션으로 설정 값을 가져온 다음에 PostConstruct로 HMAC 암호화를 진행하여 저장하는 게 더 낫다고 봐!

jang-namu replied:
<img width="690" alt="스크린샷 2023-12-24 오후 9 29 15" src="https://github.com/inu-appcenter/server-study-15.5/assets/121238128/7a2c37d3-9aa1-4614-8775-e89d63abc516">

또또 배배워갑니다..

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/security/JwtProvider.java:
```diff
@@ -0,0 +1,94 @@
+        return request.getHeader("Authorization");
+    }
+
+    public boolean validateToken(String token) {
```
Bearer 타입 체크가 들어간 건 좋은 거 같아!
지금 현재는 다른 타입이 없는데 (Basic 타입 등) 이런 상황에도 Bearer 타입 체크가 맞을까?
이 부분은 정답은 없고 형의 생각이 궁금해!

jang-namu replied:
토큰의 타입에 대해서는 Basic과 Bearer 정도밖에 모르지만,
Basic의 경우 Base64로 인코딩된 정보를 사용하고
Bearer의 경우 일바적으로 JWT 기반 인증, OAuth에서 사용한다고 알고있어서 그렇게 추가했어.

내가 이해한 것과 질문의 의도가 맞는지 잘 모르겠는데, 
이 질문은 '종류가 하나뿐인데 굳이 타입체크의 필요성이 있을까'라는 저의가 맞을까?

음.. 이 질문을 받기전엔 깊게 생각해보지 않았는데, 내 나름대로 고민하다보니 빈약하게나마 답이 나온 것 같아.
나는 이와같이 토큰값 앞에 Bearer이나 Basic을 추가로 붙이는게 일종의 프로토콜과 같은 역할을 한다고 이해했어.
그래서 지금 당장은 빼도 잘 동작하겠지만, 프론트엔드와 백엔드에서의 규약이라는 측면에서 포함시키는 게 좋지 않을까 고민해봤어.

Juser0 replied:
> 토큰의 타입에 대해서는 Basic과 Bearer 정도밖에 모르지만, Basic의 경우 Base64로 인코딩된 정보를 사용하고 Bearer의 경우 일바적으로 JWT 기반 인증, OAuth에서 사용한다고 알고있어서 그렇게 추가했어.
> 
> 내가 이해한 것과 질문의 의도가 맞는지 잘 모르겠는데, 이 질문은 '종류가 하나뿐인데 굳이 타입체크의 필요성이 있을까'라는 저의가 맞을까?
> 
> 음.. 이 질문을 받기전엔 깊게 생각해보지 않았는데, 내 나름대로 고민하다보니 빈약하게나마 답이 나온 것 같아. 나는 이와같이 토큰값 앞에 Bearer이나 Basic을 추가로 붙이는게 일종의 프로토콜과 같은 역할을 한다고 이해했어. 그래서 지금 당장은 빼도 잘 동작하겠지만, 프론트엔드와 백엔드에서의 규약이라는 측면에서 포함시키는 게 좋지 않을까 고민해봤어.

내 질문의 의도는 그게 맞아!
나도 형의 생각이 맞다고 생각해!

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/sign/UserAccountController.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    // url이 애매하다.
+    @PostMapping(value = "/login")
```
로그인의 경우에는 로그인이 맞다고 생각해!
sign-in, sign-up / login, register 등 여러 방안이 있는데 login도 괜찮다 보여

Juser0 commented on contents/todoListAPI/ch/todo/src/main/java/com/example/todo/sign/UserAccountService.java:
```diff
@@ -0,0 +1,75 @@
+
+
+    /*
+        예외처리 필요.
```
그러면 여기서 예외처리에 대한 부분은 어떤게 있을까?

jang-namu committed:
[Chore] 커밋 정리

jang-namu committed:
[Fix] Authority - user 필드 JsonIgnore 어노테이션 추가
양방향 순환 참조문제 해결

Juser0 reviewed (commented):
수고했어!

지금 Swagger에서 테스트가 가능한 지 모르겠네
그리고 Token에서 로그인 한 사용자의 정보를 가져올 수 있으면 Controller Method에서 식별자 값을 요구하는 parameter를 삭제할 수 있지 않을까?
그리고 Authority를 따로 빼는 게 맞을까 잘 모르겠네

jang-namu committed:
[Chore] Local Develop environmenet update

jang-namu committed:
[Chore] 주석 추가 및 module.iml 변경

jang-namu committed:
[Comment] 불필요한 주석 삭제

jang-namu committed:
[Feat] Swagger에 Spring Security(jwt) 적용

jang-namu committed:
[Fix] Spring Security 추가 후 LazyInitializationException 해결
원인: UserDetails의 getAuthorities()에서 User의 roles를 가져올 때,
User 엔티티를 관리하던 영속성 컨텍스트가 이미 닫힌 상태이므로 Proxy 객체인 roles는
Lazily Fetch가 불가능한 상태가 되어 예외발생
해결: @Transactional 어노테이션을 getAuthentication() 메소드 범위에 추가.
=> 해당 기간동안 영속성 컨텍스트가 유지되므로 Lazily fetch가 가능.

jang-namu committed:
[Comment] @Transactional 사용과 JPQL 사용에 대한 고민
